src/rust/kafka/src/lib.rs

```diff
@@ -0,0 +1,279 @@
+use std::{
+    fmt::Display,
+    marker::PhantomData,
+};
+
+use futures::{
+    stream::{
+        Stream,
+        StreamExt,
+    },
+    FutureExt,
+    TryFutureExt,
+    TryStreamExt,
+};
+use rdkafka::{
+    config::ClientConfig,
+    consumer::stream_consumer::StreamConsumer,
+    error::KafkaError,
+    producer::{
+        FutureProducer,
+        FutureRecord,
+    },
+    util::Timeout,
+    Message,
+};
+use rust_proto_new::{
+    SerDe,
+    SerDeError,
+};
+use thiserror::Error;
+
+//
+// Kafka configurations
+//
+
+#[derive(Error, Debug)]
+pub enum ConfigurationError {
+    #[error("failed to retrieve value from environment variable {0}")]
+    EnvironmentError(#[from] std::env::VarError),
+
+    #[error("failed to construct kafka producer {0}")]
+    ProducerCreateFailed(KafkaError),
+
+    #[error("failed to construct kafka consumer {0}")]
+    ConsumerCreateFailed(KafkaError),
+
+    #[error("failed to subscribe kafka consumer {0}")]
+    SubscriptionFailed(KafkaError),
+}
+
+fn configure() -> Result<ClientConfig, std::env::VarError> {
+    let bootstrap_servers = std::env::var("KAFKA_BOOTSTRAP_SERVERS")?;
+    if bootstrap_servers.starts_with("SASL_SSL") {
+        // running in aws w/ ccloud
+        // https://docs.confluent.io/cloud/current/client-apps/config-client.html#librdkafka-based-c-clients
+        Ok(ClientConfig::new()
+            .set("bootstrap.servers", bootstrap_servers)
+            .set("security.protocol", "SASL_SSL")
+            .set("sasl.mechanisms", "PLAIN")
+            .set("sasl.username", std::env::var("KAFKA_SASL_USERNAME")?)
+            .set("sasl.password", std::env::var("KAFKA_SASL_PASSWORD")?)
+            .set("broker.address.ttl", "30000")
+            .set("api.version.request", "true")
+            .set("api.version.fallback.ms", "0")
+            .set("broker.version.fallback", "0.10.0.0")
+            .to_owned())
+    } else {
+        // running locally
+        Ok(ClientConfig::new()
+            .set("bootstrap.servers", bootstrap_servers)
+            .set("security.protocol", "PLAINTEXT")
+            .to_owned())
+    }
+}
+
+//
+// Producer
+//
+
+fn producer() -> Result<FutureProducer, ConfigurationError> {
+    configure()?
+        .set("acks", "all")
+        .create()
+        .map_err(|e| ConfigurationError::ProducerCreateFailed(e))
+}
+
+#[derive(Error, Debug)]
+pub enum ProducerError {
+    #[error("failed to serialize message {0}")]
+    SerializationError(#[from] SerDeError),
+
+    #[error("failed to deliver message to kafka {0}")]
+    KafkaError(#[from] KafkaError),
+}
+
+pub struct Producer<T>
+where
+    T: SerDe,
+{
+    producer: FutureProducer,
+    topic: String,
+    _t: PhantomData<T>,
+}
+
+impl<T: SerDe> Clone for Producer<T> {
```

Comment:
I tried a `derive` and it broke for some reason, I'll see if I can reproduce the error.

Reply:
This is what happens when I try to `[derive(Clone)]` on `Producer<T>`: 
```
error[E0599]: the method `clone` exists for struct `Producer<U>`, but its trait bounds were not satisfied
   --> kafka/src/lib.rs:274:22
    |
97  | / pub struct Producer<T>
98  | | where
99  | |     T: SerDe,
100 | | {
...   |
103 | |     _t: PhantomData<T>,
104 | | }
    | | -
    | | |
    | |_method `clone` not found for this
    |   doesn't satisfy `Producer<U>: Clone`
...
274 |                       .clone()
    |                        ^^^^^ method cannot be called on `Producer<U>` due to unsatisfied trait bounds
    |
note: the following trait bounds were not satisfied because of the requirements of the implementation of `Clone` for `_`:
      `U: Clone`
   --> kafka/src/lib.rs:96:10
    |
96  | #[derive(Clone)]
    |          ^^^^^
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following trait defines an item `clone`, perhaps you need to implement it:
            candidate #1: `Clone`
    = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)```

Reply:
Yep, this is a known issue. https://github.com/rust-lang/rust/issues/41481

You'll find manual impls of Clone in Grapl for this reason.

Reply:
Hmm.. yeah and all the `SerDe` types _already_ implement `Clone` because they have to for tests... So I can really push the `Clone` into the `SerDe` definition itself I think. Will try that.